### PR TITLE
Consistently handle keys not found

### PR DESCRIPTION
**Pull request by jcazevedo**
This PR adds a more consistent handling of missing keys, avoiding `ConfigConverts` to have to deal with `null` in cases where a key is not found. This also disallows custom `ConfigConvert` implementations to operate on missing keys, as I think it is error-prone. For example, the following was previously allowed:

```scala
case class Bar(v: Int)
case class Foo(bar: Bar)

implicit val conv = new ConfigConvert[Bar] {
  def from(cv: ConfigValue) = Success(Bar(4))
  def to(b: Bar) = ???
}

pureconfig.loadConfig[Foo](ConfigFactory.empty())
// Would return Success(Foo(Bar(4)))
```

In this case I was expecting a `KeyNotFoundException("bar")`, regardless of the implementation of `ConfigConvert[Bar]`.

### CODE REVIEW

**Comment by melrief** [comment]
I like the PR but I have a question: do you think nobody will ever want to create a `ConfigConvert` that accepts also missing keys? This PR removes the possibility and I'm not sure this is what we want. `Option` is not a special data type and I can think about other datatypes for which "key not found" would result into one of the possible values. Do we want to have only `Option` as special datatype that can be encoded/decoded from/to `null` and not provide this possibility to the user?

File: core/src/main/scala/pureconfig/ConfigConvert.scala
Comment (leifwickland) [inline core/src/main/scala/pureconfig/ConfigConvert.scala]: It seems like the more open approach to this would be to introduce a trait, e.g. `AllowMissingConfigConvert` or `AllowNullConfigConvert`, which is used for the type check on `ConfigConvert.scala:121`.
Pureconfig would ship with `OptionConfigConvert` as the only child of that trait.

**Review by leifwickland — APPROVED** [review]
In general I like the direction you're heading with this. Thanks for the PR!

File: core/src/main/scala/pureconfig/ConfigConvert.scala
Comment (melrief) [reply core/src/main/scala/pureconfig/ConfigConvert.scala]: I like this idea, what do you think @jcazevedo ?
FYI an alternative could be to break the API again and pass an `Option[ConfigValue]` to `.from` to help the user understand that a key could be not found. I'm not sure about this approach because in most of the cases I expect a missing key to be an error, but it's a possibility.

Comment (jcazevedo) [reply core/src/main/scala/pureconfig/ConfigConvert.scala]: I like that idea, and I've implemented it in 971c2c7. I named the trait `AllowMissingKey`, though. I don't think it is necessary to put `ConfigConvert` in the name, as it will already be mixed into one, and I think `AllowMissingKey` is clearer than `AllowMissing` or `AllowNull`, but I'm open to change it.

Comment (jcazevedo) [reply core/src/main/scala/pureconfig/ConfigConvert.scala]: Even though handling `null` isn't ideal, I don't like having the `from` method receiving an `Option[ConfigValue]`, as most of the time custom implementations will only be handling `Some` values, but would still have to do something about `None`. 

One alternative would be for `ConfigConvert` to have a `onMissingKey: Option[T]` method that would be `None` by default. That would allow customizing the behavior for missing keys while not breaking the API and not having the `from` method called with `null`.

File: README.md
Comment (leifwickland) [inline README.md]: Nitpick: I'd move the "s" outside the backticks, i.e. `ConfigConvert`s.
Ditto below in a few places.
Feel free to ignore.

File: README.md
Comment (leifwickland) [inline README.md:212]: Kudos on the excellent documentation.

**Review by leifwickland — APPROVED** [review]
Thanks for doing this, @jcazevedo!

File: README.md
Comment (jcazevedo) [reply README.md]: Fixed in 156ea57.

**Comment by melrief** [comment]
I like this a lot @jcazevedo, thanks for the work. Imho ready to merge

**Review by melrief — APPROVED** [review]

**Comment by melrief** [comment]
@ruippeixotog me and @leifwickland have approved the PR, if you can have a look and you agree, we can merge this.

**Review by ruippeixotog — APPROVED** [review]
LGTM too. I'm not particularly fond of magic marker traits to customize the core library behavior, but it is a better approach than having special treatment for `OptionConfigConvert`. Other than that, it's very nice to have `ConfigConvert` implementations not having to worry with missing configs!

**Comment by melrief** [comment]
@jcazevedo do we want to release `0.5.1` right now with only this change or can it wait until #66 is merged?

**Comment by jcazevedo** [comment]
Let's wait until #66 is merged. I'll have a look at it later today.

**Pull request merged**
